feat: Amplify command Hooks

xo28122000 opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#pull-requests
-->


#### Description of changes
<!--
Thank you for your Pull Request! Please provide a description above and review
the requirements below.
-->
This PR is to add the support for amplify Command hooks. 
Runtime hooks allows customers to write scripts that would be executed by the CLI at various events. An event for this release, has a pre | post prefix followed by the command that is being executed (ex: add) and an optional sub-command that is executed (ex: auth). 

#### Issue #, if available
<!-- Also, please reference any associated PRs for documentation updates. -->
RFC for this feature: #7633 


#### Description of how you validated changes
Through manual testing and writing unit tests to verify the consistency in the following releases. 


#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [x] Relevant documentation is changed or added (and PR referenced)


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

xo28122000 committed:
feat: hooks on amplify init

xo28122000 committed:
feat: hooks on amplify init

xo28122000 committed:
fix: modified resource, fix path

xo28122000 committed:
fix: modified resource, fix path

xo28122000 committed:
fix: added hook creation in generateLocalRuntimeFiles

xo28122000 committed:
fix: added hook creation in generateLocalRuntimeFiles

xo28122000 committed:
refactor: changing hooks placement and behaviour on init

xo28122000 committed:
refactor: changing hooks placement and behaviour on init

xo28122000 committed:
feat: hooks - amplify pull - attachBackend

xo28122000 committed:
feat: hooks pull

ammarkarachi commented on packages/amplify-cli-core/src/state-manager/stateManager.ts:
```diff
@@ -236,6 +236,21 @@ export class StateManager {
+      default: {},
+    };
+
+    return this.getData<$TSAny>(filePath, mergedOptions);
```
Can you create a type for this?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+
+const S3_HOOKS_DIRECTORY = 'hooks/';
+
+export async function uploadHooksDirectory(context: $TSContext): Promise<string[]> {
```
Why are you not zipping the hooks directory?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+  // TODO: delete s3 hooks direcotry
+
+  if (!fs.existsSync(hooksDirectoryPath)) {
+    return [];
```
Do you do something with the return value? Shouldn't it be void?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+  return filesSuccessfullyUploaded;
+}
+
+export async function downloadHooks(
```
Shouldn't this return a path?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+
+export async function pullHooks(context: $TSContext): Promise<void> {
+  /**
+   * pulls hooks directory from S3 and places in amplify project.
```
it's S3 or Amplify?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+// general utility functions:
+function placeFile(filePath: string, data: $TSAny): void {
+  /**
+   * quietly ensures a file esists else creates one and writes the data
```
exists

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.ts:
```diff
@@ -160,6 +160,27 @@ export class S3 {
   }
 
+  async getObjects(s3Params: $TSAny, envName?: string): Promise<$TSAny[]> {
+    if (!s3Params.hasOwnProperty('Bucket')) {
```
refactor with other methods too

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+    // log();
+    listHookObjects = await s3Client.listObjects(params).promise();
+  } catch (ex) {
+    // log(ex);
```
Throw this

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,256 @@
+      // log();
+      hooksFileObject = await s3Client.getObject(params).promise();
+    } catch (ex) {
+      // log(ex);
```
throw this

ammarkarachi commented on packages/amplify-cli/src/pull-backend.ts:
```diff
@@ -13,6 +13,9 @@ export async function pullBackend(context: $TSContext, inputParams: $TSAny) {
   context.exeInfo.forcePush = false;
   context.exeInfo.restoreBackend = !context.exeInfo.inputParams.amplify.noOverride;
 
+  context.print.info(`Continuing will automatically override the hooks directory.`);
```
you can safely remove this print statement

xo28122000 committed:
feat: hooks on amplify push

akshbhu commented on packages/amplify-cli-core/src/state-manager/stateManager.ts:
```diff
@@ -236,6 +236,21 @@ export class StateManager {
+    return this.getData<$TSAny>(filePath, mergedOptions);
+  };
+  setSampleHooksDir = (projectPath: string | undefined, dirSourcePath: string): void => {
+    const toDirPath = pathManager.getHooksDirPath(projectPath);
```
```suggestion
    const targetDirPath = pathManager.getHooksDirPath(projectPath);
```

akshbhu commented on packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.ts:
```diff
@@ -160,6 +160,27 @@ export class S3 {
+
+    // TODO: logs
+    //   const log = logger("downloadHooks.s3.listObjects", [params]);
+    let listObjects;
```
const?

akshbhu commented on packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.ts:
```diff
@@ -160,6 +160,27 @@ export class S3 {
+      listObjects = await this.s3.listObjects(s3Params).promise();
+      return listObjects.Contents;
+    } catch (ex) {
+      // log(ex);
```
throw error here

akshbhu commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,268 @@
+
+  const s3 = await S3.getInstance(context);
+  try {
+    await s3.deleteDirectory(projectBucket, S3_HOOKS_DIRECTORY);
```
Check if the delete directory exists?

xo28122000 committed:
feat: hooks execution

xo28122000 committed:
refactor: cr recs, feat: run post-hoooks on amplify error

xo28122000 committed:
test: added hooks exports, attached getEnvInfo function

xo28122000 committed:
test: for hooksHandler, refactor

xo28122000 committed:
test: added tests for hooksExecutioner

xo28122000 committed:
test: added tests for hooksManager

xo28122000 committed:
feat: added skiphooks, test: added tests for skiphooks

edwardfoyle commented on packages/amplify-cli/src/commands/push.ts:
```diff
@@ -38,6 +38,26 @@ async function syncCurrentCloudBackend(context: $TSContext) {
+
+  try {
+    const providerPlugins = getProviderPlugins(context);
+    const pushHooksTasks: (() => Promise<$TSAny>)[] = [];
```
```suggestion
    const pushHooksTasks: (() => Promise<void>)[] = [];
```

edwardfoyle commented on packages/amplify-cli/src/index.ts:
```diff
@@ -107,8 +109,16 @@ export async function run() {
 
     rewireDeprecatedCommands(input);
     logInput(input);
+    const hooksHandler = HooksHandler.initialize();
+    hooksHandler.setHooksEventFromInput(input);
     const context = constructContext(pluginPlatform, input);
 
+    try {
+      hooksHandler.dataParameter.amplify.environment = context.amplify.getEnvInfo()?.envName;
+    } catch (err) {
+      // do nothing
+    }
+
```
these changes should go below the ff initialization

edwardfoyle commented on packages/amplify-cli/src/index.ts:
```diff
@@ -219,6 +229,10 @@ export async function run() {
         print.info(error.stack);
       }
     }
+    await executeHooks(undefined, 'post', {
```
what does this do? shouldn't a hook event always be passed here?

edwardfoyle commented on packages/amplify-cli/src/init-steps/s9-onSuccess.ts:
```diff
@@ -177,6 +179,14 @@ function generateReadMeFile(context: $TSContext) {
 }
 
+function generateHooksSampleFiles(context: $TSContext) {
+  // TODO: delete amplify/hooks if exist
```
I'm not sure we want to do that. What if a customer pulls a sample project from github and then runs `amplify init` on it?

edwardfoyle commented on packages/amplify-cli-core/src/state-manager/pathManager.ts:
```diff
@@ -156,6 +161,18 @@ export class PathManager {
 
   getDeploymentSecrets = (): string => path.normalize(path.join(this.getDotAWSAmplifyDirPath(), PathConstants.DeploymentSecretsFileName));
 
+  getHooksDirPath = (projectPath?: string): string =>
```
I don't think it makes sense for these hook-specific paths to be part of PathManager (especially not the sample file paths) or StateManager, below.

I would recommend creating a `hookStateManager` object that can expose functions for all hook state management related operations

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/__tests__/utils/hooks-manager.test.ts:
```diff
@@ -0,0 +1,203 @@
+jest.mock('process');
```
can we mock the return values of fs instead of including all these empty test files?

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+export const S3_HOOKS_DIRECTORY = 'hooks/';
+
+export async function uploadHooksDirectory(context: $TSContext): Promise<void> {
+  /**
```
same as earlier, JS docs should be above function sig (here and everywhere else)

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+   */
+  try {
+    fs.ensureFileSync(filePath);
+    fs.writeFileSync(filePath, data);
```
writeFile will create the file if it doesn't exist by default

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+  try {
+    fs.ensureFileSync(filePath);
+    fs.writeFileSync(filePath, data);
+  } catch (ex) {
+    throw ex;
+  }
```
remove try catch

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+      }
+    }
+  }
+  return;
```
remove

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+      };
+      try {
+        await s3.uploadFile(s3Params);
+      } catch (ex) {
+        throw ex;
+      }
```
remove try catch

ammarkarachi replied:
Or add logging

edwardfoyle commented on packages/amplify-provider-awscloudformation/src/utils/hooks-manager.ts:
```diff
@@ -0,0 +1,275 @@
+    // TODO: logs
+    // const log = logger('downloadHooks.s3.getObject', [params]);
+    let hooksFileObject = null;
+    try {
+      // log();
+      hooksFileObject = await s3Client.getObject(params).promise();
+    } catch (ex) {
+      // log(ex);
+      throw ex;
+    }
```
remove try catch (and anywhere else where the error is just rethrown in the catch)

xo28122000 committed:
feat: added logs, refactor

xo28122000 committed:
refactor(cli): rebasing the branch

xo28122000 committed:
refactor: fixes and address comments

xo28122000 committed:
refactor: responding to comments, cleaning tests

xo28122000 committed:
refactor: fixes and resolving comments

xo28122000 committed:
feat: added support for --forcePush flag

xo28122000 committed:
fix: doesnt throw error on fresh init

xo28122000 committed:
fix: running post event hooks before every exit

xo28122000 committed:
fix: runs with no env info

xo28122000 committed:
refactor: api changes

xo28122000 committed:
feat: added e2e tests, fixed newline problem

ammarkarachi commented on packages/amplify-cli-core/src/hooks/hooksHandler.ts:
```diff
@@ -0,0 +1,100 @@
+import { suppportedEvents, supportedEnvEvents } from './hooksConstants';
+import { stateManager } from '../state-manager';
+
+export class HooksHandler {
```
Can you add a few lines what the class does and what it is for?

edwardfoyle replied:
> the main purpose of this class is to capture the hook event and data parameter as early as possible

This doesn't seem like it should be a singleton and doesn't seem like it should be named `HooksHandler`. Why not just a normal (non-singleton) class called `HookEvent` or something similar?

ammarkarachi commented on packages/amplify-provider-awscloudformation/src/aws-utils/aws-s3.ts:
```diff
@@ -140,7 +148,7 @@ export class S3 {
     bucketName: string,
     options: OptionalExceptFor<ListObjectVersionsOutput, 'KeyMarker' | 'VersionIdMarker'> = null,
-  ) {
+  ): Promise<Required<ObjectIdentifier>[]> {
```
This looks kinda sus `Required<ObjectIdentifier>`

ammarkarachi commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,215 @@
+    // do nothing
+  }
+
+  // merging because we want to remoe as many undeifined values as possible
```
remove

xo28122000 committed:
refactor: using amplify-prompts

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksHandler.ts:
```diff
@@ -0,0 +1,100 @@
+   * @internal
+   * private method used in unit tests to release the instance
+   */
+  public static releaseInstance = (): void => {
```
See above comment about making this a non-singleton class. In general, exposing methods just for testing is a code smell

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksTypes.ts:
```diff
@@ -0,0 +1,40 @@
+
+export type EventPrefix = 'pre' | 'post';
+
+export type HooksEvent = {
```
What do you mean "singular context"? If it's referring to a single hook, it should be `HookEvent`

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksTypes.ts:
```diff
@@ -0,0 +1,40 @@
+
+export type DataParameter = {
+  amplify: {
+    version?: string;
+    environment?: string;
+    command?: string;
+    subCommand?: string;
+    argv?: string[];
```
Regarding the "optionalness of all the above types, I would recommend making any field that will be present when we call the hook required. Internally, if we want to initialize a partial object and build it up gradually, we can use `Partial<Type>` and then assert it to `Type` when we have all the values

edwardfoyle commented on packages/amplify-cli/src/commands/version.ts:
```diff
@@ -1,9 +1,6 @@
-  const pkg = JSONUtilities.readJson<$TSAny>(path.join(__dirname, '..', '..', 'package.json'));
-
-  context.print.info(pkg.version);
+  context.print.info(getAmplifyVersion());
```
printer.info

edwardfoyle commented on packages/amplify-cli/src/extensions/amplify-helpers/get-amplify-version.ts:
```diff
@@ -0,0 +1,7 @@
+import * as path from 'path';
+import { JSONUtilities, $TSAny } from 'amplify-cli-core';
+
+export const getAmplifyVersion = (): string => {
```
There's going to be duplicate logic once this is merged: https://github.com/aws-amplify/amplify-cli/pull/7834/files#diff-a536d7e42a68805a2355f24a1186275d46f210bad20c4b5465ab3592ad277da4R63

edwardfoyle commented on packages/amplify-cli/resources/sampleHooksDirectory/Readme.md:
```diff
@@ -0,0 +1,7 @@
+# Runtime Hooks
```
path should be `resources/sample-hooks`

edwardfoyle commented on packages/amplify-e2e-core/src/init/amplifyPush.ts:
```diff
@@ -246,3 +246,17 @@ export function amplifyPushMissingFuncSecret(cwd: string, newSecretValue: string
+  return new Promise((resolve, reject) => {
+    spawn(getCLIPath(testingWithLatestCodebase), ['push'], { cwd, stripColors: true, noOutputTimeout: pushTimeoutMS })
+      .wait('No changes detected')
+      .run((err: Error) => {
+        if (!err) {
+          resolve();
+        } else {
+          reject(err);
+        }
+      });
```
```suggestion
      .run((err: Error) => err ? reject(err) : resolve());
```

edwardfoyle commented on packages/amplify-e2e-core/src/utils/hooks.ts:
```diff
@@ -0,0 +1,5 @@
+import * as path from 'path';
+
+export const getHooksDirPath = (projRoot: string): string => {
```
isn't this already in pathManager?

xo28122000 replied:
its in the e2e-core for not using path manager and making sure the path in path manager is not changed.

edwardfoyle reviewed (commented):
Overall looking pretty good!
My one general comment is be careful about using "hook" vs "hooks". I think in almost all cases the objects and functions are dealing with one hook at a time so the names should be singular.

xo28122000 commented on packages/amplify-cli-core/src/hooks/hooksConstants.ts:
```diff
@@ -0,0 +1,70 @@
+export const hooksFileSeperator = '-';
+
+export const suppportedEvents: Record<HooksVerb, Set<HooksNoun>> = {
+  add: new Set([
```
okay I'll refactor it into an array

xo28122000 commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,215 @@
+import { EOL } from 'os';
+const logger = getLogger('amplify-cli-core', 'hooks/hooksExecutioner.ts');
+
+export const executeHooks = async (
```
So this would execute multiple hooks, pre-add, pre-add-auth, in case of forcePush, it would execute push hooks as well

xo28122000 commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,215 @@
+  }
+};
+
+const exec = async (
```
I'll refactor the name

xo28122000 committed:
refactor: resolving reviews

Removed try catch, refactored to use printer, Changed filename hooksMeta.ts and hooksMeta.test.ts,
Changed classname hooksMeta, Changed to getInstance() in hooksMeta

xo28122000 committed:
refactor: sleak implementation

Co-authored-by: Edward Foyle <foyleef@amazon.com>

xo28122000 committed:
Refactor: Apply suggestions from code review

Co-authored-by: Edward Foyle <foyleef@amazon.com>

xo28122000 committed:
refactor: resolving reviews

Changed suppportedEvents and supportedEnvEvents to be arrays, changed type to be required where
possible

xo28122000 committed:
test: added tests for constructHooksMetaObject function

xo28122000 committed:
refactor: naming convention changes

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,185 @@
+
+  const { commandHookFileMeta, subCommandHookFileMeta } = getHookFileMetas(hooksDirPath, hooksMeta.getHookEvent(), hooksConfig);
+
+  let executionQueue = [commandHookFileMeta, subCommandHookFileMeta];
```
nit: const

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,185 @@
+    executionQueue.push(commandHookFileMeta);
+  }
+
+  for (const execFileMeta of executionQueue) {
+    if (execFileMeta) {
+      const runtime = getRuntime(execFileMeta, hooksConfig);
+      if (runtime) {
+        await execHelper(runtime, execFileMeta, hooksMeta.getDataParameter(), hooksMeta.getErrorParameter());
+      }
+    }
+  }
+};
```
nit:
```suggestion
  for (const execFileMeta of executionQueue) {
    if (!execFileMeta) {
      continue;
    }
    const runtime = getRuntime(execFileMeta, hooksConfig);
    if (!runtime) {
      continue!;
    }
    await execHelper(runtime, execFileMeta, hooksMeta.getDataParameter(), hooksMeta.getErrorParameter());
};
```

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+
+export class HooksMeta {
+  private static instance?: HooksMeta;
+  private HookEvent: Partial<HookEvent>;
```
nit:
```suggestion
  private hookEvent: Partial<HookEvent>;
```

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+  private static instance?: HooksMeta;
+  private HookEvent: Partial<HookEvent>;
+  private dataParameter: DataParameter;
+  private errorParameter: ErrorParameter | undefined;
```
nit:
```suggestion
  private errorParameter?: ErrorParameter;
```

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+    this.HookEvent.subCommand = subCommand;
+  }
+
+  public setEventPrefix(prefix?: string): void {
```
should the input type here be EventPrefix?

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+        this.HookEvent.subCommand = subCommand;
+      }
+    }
+    this.HookEvent.forcePush = (input?.options?.forcePush && this.HookEvent.command !== 'push') ?? false;
```
in what cases is this true?

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksExecutor.ts:
```diff
@@ -0,0 +1,185 @@
+import { printer } from 'amplify-prompts';
+const logger = getLogger('amplify-cli-core', 'hooks/hooksExecutioner.ts');
+
+export const executeHooks = async (hooksMeta: HooksMeta): Promise<void> => {
```
Nice! this is much cleaner :)

edwardfoyle commented on packages/amplify-cli/src/display-banner-messages.ts:
```diff
@@ -9,6 +9,10 @@ export async function displayBannerMessages(input: Input) {
   }
   await displayLayerMigrationMessage();
+  if (skipHooks()) {
+    printer.warn('Warning: Amplify runtime hooks are disabled in the current execution environment.');
```
nit: `printer.warn` will prefix the line with ⚠️ so adding `Warning:` may be redundant

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+    return HooksMeta.instance;
+  };
+
+  public static constructHooksMetaObject = (
```
suggest renaming this to `initialize`

xo28122000 replied:
done

edwardfoyle commented on packages/amplify-cli-core/src/hooks/hooksMeta.ts:
```diff
@@ -0,0 +1,155 @@
+  private dataParameter: DataParameter;
+  private errorParameter: ErrorParameter | undefined;
+
+  public static getInstance = (HookEvent: Partial<HookEvent> = {}, dataParameter: DataParameter = { amplify: {} }): HooksMeta => {
```
is there any case where this should be called before `constructHooksMetaObject`? If not, I would suggest modifying this method to throw if not initialized. This makes it clear that `constructHooksMetaObject` (which I propose renaming to `initialize`) must be called before an instance can be returned.

My concern here is that there is some interplay between `constructHooksMetaObject` and `getInstance` that is not obvious just from the interface. If we can make it obvious that `constructHooksMetaObject` must be called first, that would be ideal

Regardless of what you decide above, it doesn't looks like the input params are supplied anywhere so I would remove them.

xo28122000 replied:
done

xo28122000 committed:
refactor: resolving comments

xo28122000 committed:
refactor: var name change

edwardfoyle reviewed (approved):
LGTM! Great work and thanks for working through the changes!

srquinn21 commented on packages/amplify-cli-core/src/hooks/hooksConstants.ts:
```diff
@@ -0,0 +1,46 @@
+
+export const defaultSupportedExt: HookExtensions = { js: { runtime: 'node' }, sh: { runtime: 'bash' } };
+
+export const skipHooksFileName: string = 'AMIPLIFY-HOOKS-SKIP';
```
NIt: this is not used and should be deleted.

srquinn21 commented on packages/amplify-cli-core/src/hooks/hooksConstants.ts:
```diff
@@ -0,0 +1,46 @@
+export const defaultSupportedExt: HookExtensions = { js: { runtime: 'node' }, sh: { runtime: 'bash' } };
+
+export const skipHooksFileName: string = 'AMIPLIFY-HOOKS-SKIP';
+export const skipHooksFilePath: string = join(homedir(), 'opt', 'amazon');
```
We should explicitly use `/opt/amazon` here — the run-as-user does not have a home directory whereas `/opt/amazon` is always present.

akshbhu committed:
chore: address comments and remaned runtime hooks in display

srquinn21 reviewed (approved):
LGTM
